src/ChannelServer/World/Entities/Components/Buffs.cs

```diff
@@ -0,0 +1,247 @@
+
+namespace Melia.Channel.World
+{
+	/// <summary>
+	/// Character buffs.
+	/// </summary>
+	public class Buffs : IUpdatableComponent
+	{
+		private readonly Dictionary<BuffId, Buff> _buffs = new Dictionary<BuffId, Buff>();
+
+		/// <summary>
+		/// The owner of this object.
+		/// </summary>
+		public ICombatEntity Owner { get; }
+
+		/// <summary>
+		/// Creates new instance for character.
+		/// </summary>
+		/// <param name="entity"></param>
+		public Buffs(ICombatEntity entity)
+		{
+			this.Owner = entity;
+		}
+
+		/// <summary>
+		/// Returns the amount of buffs in the collection.
+		/// </summary>
+		public int Count { get { lock (_buffs) return _buffs.Count; } }
+
+		/// <summary>
+		/// Adds given without updating the client. Replaces existing
+		/// buffs.
+		/// </summary>
+		/// <param name="buff"></param>
+		public void AddSilent(Buff buff)
+		{
+			lock (_buffs)
+				_buffs[buff.Id] = buff;
+		}
+
+		/// <summary>
+		/// Adds given buff and updates the client.
+		/// </summary>
+		/// <param name="buff"></param>
+		public void Add(Buff buff)
+		{
+			buff.IncreaseOverbuff();
+			this.AddSilent(buff);
+			var handler = ChannelServer.Instance.BuffHandlers.GetBuff(buff.Id);
+			handler.OnStart(buff);
+			Send.ZC_BUFF_ADD(this.Owner, buff);
+		}
+
+		/// <summary>
+		/// Removes buff with given id, returns false if it
+		/// didn't exist. Doesn't update the client.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <returns></returns>
+		public bool RemoveSilent(BuffId buffId)
+		{
+			lock (_buffs)
+				return _buffs.Remove(buffId);
+		}
+
+		/// <summary>
+		/// Removes buff with given id, returns false if it
+		/// didn't exist. Updates the client on success.
+		/// </summary>
+		/// <param name="buffId"></param>
+		/// <returns></returns>
+		public bool Remove(BuffId buffId)
+		{
+			var buff = this.Get(buffId);
+			if (buff != null)
+			{
+				this.Remove(buff);
+			}
+			return false;
+		}
+
+		/// <summary>
+		/// Removes buff, returns false if it didn't exist. 
+		/// Updates the client on success.
+		/// </summary>
+		/// <param name="buff"></param>
+		/// <returns></returns>
+		public bool Remove(Buff buff)
+		{
+			var isRemoved = false;
+			lock (_buffs)
+			{
+				isRemoved = _buffs.Remove(buff.Id);
+			}
+			if (isRemoved)
+			{
+				var handler = ChannelServer.Instance.BuffHandlers.GetBuff(buff.Id);
+				handler.OnEnd(buff);
+				Send.ZC_BUFF_REMOVE(this.Owner, buff);
+			}
+			return isRemoved;
+		}
+
+		/// <summary>
+		/// Add Buff(s) or Update Buff(s) if it exists
+		/// with a given id.
+		/// </summary>
+		/// <param name="buffs"></param>
+		public void AddOrUpdate(params Buff[] buffs)
+		{
+			foreach (var buff in buffs)
+			{
+				if (!this.Has(buff.Id))
+					this.Add(buff);
+				else
+					this.Update(buff.Id);
+			}
+		}
+
+		/// <summary>
+		/// Update a buff if it exists using the buff id.
+		/// </summary>
+		/// <param name="buffId"></param>
+		public void Update(BuffId buffId)
```

Comment:
Wouldn't private be more appropriate for this method? Or is other code supposed to call this?

Reply:
Both Add and Update define behavior for buffs, you are correct we can just set them to private and allow access to the AddOrUpdate function instead.